### PR TITLE
Istio: add gateway-virtualservice link

### PR DESCRIPTION
**Pull request by tomerlange**
This PR adds links between gateways and virtualservices.
The relevant tests are `TestBookInfoScenario` and `TestIstioGatewayVirtualServiceScenario`, which can be found at `tests/istio_test.go`.

### CODE REVIEW

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

File: tests/istio_test.go [hunk 1]
```diff
@@ -104,12 +104,38 @@ func TestIstioDestinationRuleServiceScenario(t *testing.T) {
...
+}
+
 func TestBookInfoScenario(t *testing.T) {
-	bookinfo := "WITH_ISTIO=true ./bookinfo/bookinfo.sh"
+	bookinfo := "./bookinfo/bookinfo.sh"
 	testRunner(
 		t,
 		[]Cmd{
-			{bookinfo + " start", true},
+			{bookinfo + " start WITH_ISTIO=true", true},
```
Comment (hunchback) [inline tests/istio_test.go]: Note that within the script WITH_ISTIO=false, can you please re-check

```
[vagrant@dev tests]$ ./bookinfo/bookinfo.sh start WITH_ISTIO=true
export NAMESPACE=default
export WITH_ISTIO=false
export WITH_NETPOL=false
export WITH_IBMCLOUD=false
export SRVNAME=productpage
Error from server (AlreadyExists): namespaces "default" already exists
service "details" created
deployment.extensions "details-v1" created
service "ratings" created
deployment.extensions "ratings-v1" created
service "reviews" created
deployment.extensions "reviews-v1" created
deployment.extensions "reviews-v2" created
deployment.extensions "reviews-v3" created
service "productpage" created
deployment.extensions "productpage-v1" created
ingress.extensions "gateway" created
[vagrant@dev tests]$
```

File: tests/istio_test.go [hunk 2]
```diff
@@ -104,12 +104,38 @@ func TestIstioDestinationRuleServiceScenario(t *testing.T) {
...
+				}
+				return nil
+			},
+		},
+	)
+}
+
 func TestBookInfoScenario(t *testing.T) {
-	bookinfo := "WITH_ISTIO=true ./bookinfo/bookinfo.sh"
+	bookinfo := "./bookinfo/bookinfo.sh"
```
Comment (hunchback) [inline tests/istio_test.go]: I believe this breaks both start and stop - please revert.

File: tests/istio/gateway-virtualservice.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  hosts:
+  - reviews.prod.svc.cluster.local
+  gateways:
+  - skydive-test-gateway-virtualservice
+  http:
+  - route:
+    - destination:
+        host: reviews.prod.svc.cluster.local
+        subset: v1
+
```
Comment (hunchback) [inline tests/istio/gateway-virtualservice.yaml]: Please remove the empty line

File: topology/probes/istio/gateway.go [hunk 4]
```diff
@@ -44,3 +45,19 @@ func (h *gatewayHandler) Dump(obj interface{}) string {
 func newGatewayProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
 	return k8s.NewResourceCache(client.(*kiali.IstioClient).GetIstioNetworkingApi(), &kiali.Gateway{}, "gateways", g, &gatewayHandler{})
 }
+
+func gatewayVirtualServiceAreLinked(a, b interface{}) bool {
+	gateway := a.(*kiali.Gateway)
+	vs := b.(*kiali.VirtualService)
+	vsGateways := vs.Spec["gateways"].([]interface{})
```
Comment (hunchback) [inline topology/probes/istio/gateway.go]: You can't not "gateways" exists - please verify it does before using it - see how I created a virtualservice which is missing the gateways field:

```
[vagrant@dev istio]$ git diff
diff --git a/tests/istio/gateway-virtualservice.yaml b/tests/istio/gateway-virtualservice.yaml
index 874827ad..63213d04 100644
--- a/tests/istio/gateway-virtualservice.yaml
+++ b/tests/istio/gateway-virtualservice.yaml
@@ -20,11 +20,10 @@ metadata:
 spec:
   hosts:
   - reviews.prod.svc.cluster.local
-  gateways:
-  - skydive-test-gateway-virtualservice
+  #gateways:
+  #- skydive-test-gateway-virtualservice
   http:
   - route:
     - destination:
         host: reviews.prod.svc.cluster.local
         subset: v1
-
[vagrant@dev istio]$ kubectl describe gateway/skydive-test-gateway-virtualservice
Name:         skydive-test-gateway-virtualservice
Namespace:    default
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"networking.istio.io/v1alpha3","kind":"Gateway","metadata":{"annotations":{},"name":"skydive-test-gateway-virtualservice","namespace":"de...
API Version:  networking.istio.io/v1alpha3
Kind:         Gateway
Metadata:
  Cluster Name:
  Creation Timestamp:  2019-02-24T06:59:00Z
  Generation:          1
  Resource Version:    247461
  Self Link:           /apis/networking.istio.io/v1alpha3/namespaces/default/gateways/skydive-test-gateway-virtualservice
  UID:                 a9af92dc-3801-11e9-a066-5254001c08b7
Spec:
  Selector:
    Istio:  ingressgateway
  Servers:
    Hosts:
      *
    Port:
      Name:      http
      Number:    80
      Protocol:  HTTP
Events:          <none>
[vagrant@dev istio]$
```

**Comment by hunchback** [comment]
image

does not show the link connecting the gateway to virtualservice - I believe this is not consistent in all runs - kindly re-check

**Comment by hunchback** [comment]
> image
> 
> does not show the link connecting the gateway to virtualservice - I believe this is not consistent in all runs - kindly re-check

I have checked this and this is not due to a bug in your code - rather this is a general issue effecting also other k8s resources (and possibly physical resources). so given that this issue will be solved outside the context of this PR.

**Pull request merged**